source/common.c

```diff
@@ -28,34 +33,114 @@
 #    pragma warning(disable : 4100)
 #endif
 
-static void *s_default_malloc(struct aws_allocator *allocator, size_t size) {
+#ifdef AWS_DEFAULT_ALLOCATOR_LEAKCHECK
+struct aws_allocation {
+    const char *file;
+    int line;
+    size_t size;
+
+    struct aws_allocation *next;
+    struct aws_allocation *prev;
+};
+
+struct aws_mutex s_mutex = AWS_MUTEX_INIT;
+struct aws_allocation *s_head(void) {
+    static struct aws_allocation head;
+    static bool init;
+
+    if (!init) {
+        head.next = &head;
+        head.prev = &head;
+        init = 1;
+    }
+
+    return &head;
+}
+
+static void *s_hook_leakchecker(void *mem, size_t size, const char *file, int line) {
+    if (!mem) {
+        return mem;
+    }
+
+    struct aws_allocation* allocation = (struct aws_allocation *)mem;
+    allocation->file = file;
+    allocation->line = line;
+    allocation->size = size;
+    allocation->next = allocation;
+    allocation->prev = allocation;
+
+    aws_mutex_lock(&s_mutex);
+    struct aws_allocation *head = s_head();
+    allocation->prev = head;
+    allocation->next = head->next;
+    head->next->prev = allocation;
+    head->next = allocation;
+    aws_mutex_unlock(&s_mutex);
+
+    mem = (void *)(allocation + 1);
```

Comment:
The alignment of this memory may be incorrect. For example, on x86_64, your struct will be `5 * 8` bytes, with 8 byte alignment, but there are certain SIMD operations which require 16-byte or 32-byte alignment.

You'll either need to add some platform-specific minimum alignment values, or use something like a hash table instead.